### ISSUE TITLE
Provide Flat Modelica output for structural analysis algorithms

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#9694]
Ernesto Kofman's group works on advanced array-preserving algorithms for structural analysis. They have a micro-Modelica parser that can handle Modelica models as inputs, but only handles a small subset of the language. Thy could substantially improve the scope of that, and also start prototyping the inclusion of those algorithms in the new backend, by taking the flat Modelica output of OMC of models that they cannot directly handle with their tool.

There is however one issue there: their tools are focused on set-based E-V graphs, so they accept Modelica models that basically only contain variable and array declarations, scalar and array equations. Records and functions are not handled. This prevents, e.g., handling models using Complex types, that are very commonplace for large model of electrical power systems, and also handling models with simple functions that can be inlined.

Inlining and handling of simple record equations is currently done in the backend. In order to make it possible to use Flat Modelica input for those tools, we should make it possible to optionally activate them in the front end, before flat Modelica output takes place.

I will add some MWE here to this ticket to show what kind of functionality is needed. @perost, can you please estimate how much effort it is to implement that in the NF, or to port that from the NB?

**Comment by casella** [comment]
Consider this MWE:
```modelica
package TestInlining
  function f
    input Real x;
    input Real y;
    output Real z;
  algorithm
    z := x^2 + y^2 - 1;
  annotation(Inline = true);
  end f;
  
  model M
    Real a,b;
  equation
    b = time;
    der(a) = f(a, b);
  end M;
  
  model S
    M m[10];
  end S;
end TestInlining;
```
Calling `instantiateModel(TestInlining.M)` with flags `-f -d=nonfScalarize,combineSubscripts,printRecordTypes, --newBackend --showStructuralAnnotations` gives
```modelica
model 'TestInlining.M'
function 'TestInlining.f'
  input Real 'x';
  input Real 'y';
  output Real 'z';
algorithm
  'z' := 'x' ^ 2.0 + 'y' ^ 2.0 - 1.0;
  annotation(Inline = true);
end 'TestInlining.f';

  Real 'a';
  Real 'b';
equation
  'b' = time;
  der('a') = 'TestInlining.f'('a', 'b');
end 'TestInlining.M';
```
It should instead give
```modelica
model 'TestInlining.M'
  Real 'a';
  Real 'b';
equation
  'b' = time;
  der('a') = 'a'^2.0 + 'b'^2.0 - 1;
end 'TestInlining.M';
```
while flattening `S` should give
```modelica
model 'TestInlining.S'
  Real[10] 'm.b';
  Real[10] 'm.a';
equation
  'm.b'[:] = time;
  for '$i1' in 1:10 loop
    der('m.a'['$i1']) = 'm.a'['$i1']^2.0 + 'm.b'['$i1']^2.0 - 1;
  end for;
end 'TestInlining.S';
```
Inlining should be limited to functions that have one single assigment in the algorithm, so that the implementation is trivially obtain by subsitution of the formal parameters with the actual parameters.

**Comment by perost** [comment]
This actually already works if you instead use `annotation(__OpenModelica_EarlyInline = true)`. We use it for some of the builtin functions.

**Comment by casella** [comment]
For record handling, the idea is to first inline the functions, and then split the records and the record equations into their individual components. For example:
```modelica
package TestComplex
  model M
    Complex x, y, z;
  equation
    x = Complex(2.0, 3.0);
    y = x;
    x + y = z;
  end M;
  
  model S
    M m[10];
  end S;
annotation(uses(Complex(version="4.0.0")));
end TestComplex;
```
Flattening `TestComplex.M` with the flags mentioned above currently gives:
```modelica
model 'TestComplex.M'
function 'Complex.\'+\'' "Add two complex numbers"
  input 'Complex' 'c1';
  input 'Complex' 'c2';
  output 'Complex' 'c3';
algorithm
  'c3' := 'Complex.\'constructor\'.fromReal'('c1'.'re' + 'c2'.'re', 'c1'.'im' + 'c2'.'im');
  annotation(Inline = true);
end 'Complex.\'+\'';

function 'Complex.\'constructor\'.fromReal' "Construct Complex from Real"
  input Real 're';
  input Real 'im' = 0.0;
  output 'Complex' 'result';
algorithm
  annotation(Inline = true);
end 'Complex.\'constructor\'.fromReal';

record 'Complex'
  Real 're';
  Real 'im';
end 'Complex';

  'Complex' 'x';
  'Complex' 'y';
  'Complex' 'z';
equation
  'x' = 'Complex'(2.0, 3.0);
  'y' = 'x';
  'Complex.\'+\''('x', 'y') = 'z';
end 'TestComplex.M';
```
Inlining the `Complex.'+'` function should transform the equation section to
```modelica
equation
  'x' = 'Complex'(2.0, 3.0);
  'y' = 'x';
  'Complex'('x'.'re' + 'y'.'re', 'x'.'im' + 'y'.'im') = 'z';
```
Then, records and record equations can be split into their elements, which will be promoted as first-class-citizen variables

```modelica
model 'TestComplex.M'
  Real 'x.re';
  Real 'x.im';
  Real 'y.re';
  Real 'y.im';
  Real 'z.re';
  Real 'z.im';
equation
  'x.re' = 2.0;
  'x.im' = 3.0;
  'y.re' = 'x.re';
  'y.im' = 'y.im';
  'x.re' + 'y.re' = 'z.re';
  'x.im' + 'y.im' = 'z.im';
end 'TestComplex.M';
```
The same process applied to TextComplex.S should give something like:
```modelica
model 'TestComplex.S'
  Real[10] 'm.x.re';
  Real[10] 'm.x.im';
  Real[10] 'm.y.re';
  Real[10] 'm.y.im';
  Real[10] 'm.z.re';
  Real[10] 'm.z.im';
equation
  'm.x.re' = fill(2.0, 10);
  'm.x.im' = fill(3.0; 10);
  'm.y.re' = 'm.x.re';
  'm.y.im' = 'm.y.im';
  'm.x.re' + 'm.y.re' = 'm.z.re';
  'm.x.im' + 'm.y.im' = 'm.z.im';
end 'TestComplex.S';
```

```modelica
```

**Comment by perost** [comment]
That does not currently work, even if early inline annotations are added to `Complex`. I guess we don't consider inlining on operator functions, or perhaps we inline before we resolve operators. But we have the inlining functionality, so it's just a matter of making sure it's applied to those functions somehow.

**Comment by casella** [comment]
Adding @ernestokofman and @joaquinffernandez to the loop.

**Comment by casella** [comment]
> Also note that the default for record variables are to be split into their parts, the reason why they were not split in your example was because you used the `--newBackend` flag which disables that.
> 
> So the question now is how we should proceeed. I understand that you want the record equations to also be split into their parts. I don't think we have something like that implemented yet, but that should be easy to do. But I'm not sure how splitting record variables will interact with the new backend since I think it's designed around not doing that.

Good question. The answer is, we need to be flexible, depending on the requirements, by using flags appropriately. At the moment I see three different use cases:

1. The NF is used as a frontend for the new backend. Here we should leave inlining and record handling to the backend, which will eventually need to handle all possible combinations of arrays, records, and functions. Maybe @kabdelhak can comment on that. Once #9697 is resolved, the flags here will be `--newBackend` (which enables -d=nonfScalarize) and possibly -d=arrayConnect if we want to use that algorithm on array connects, otherwise they will be scalarized.
2. The NF is used as a frontend for the MARCO array-preserving LLVM compiler. I understand MARCO already handles operator records and inlining, so it would be fine to keep --newBackend there, at least in general. For power grid models, we can experiment both approaches, i.e. having either OMC NF or MARCO take care of operator record Complex, and compare the performance. I guess it will be similar, at the end of the day that task is no big deal. Here we need a lot of flags, also to get the flat textual output right, currently `-f -d=nonfScalarize,arrayConnect,combineSubscripts,printRecordTypes,evaluateAllParameters,vectorizeBindings --newBackend --showStructuralAnnotations` for array models and `-f -d=nonfScalarize,mergeComponents,combineSubscripts,printRecordTypes,evaluateAllParameters,vectorizeBindings --newBackend --showStructuralAnnotations` for models with many scalar instances of the same model.
3. The NF is used to provide flat models to @ernestokofman's group to test array-preserving structural analysis algorithms. For this purpose, it is necessary to get rid of records, because they don't handle them (yet). Functions are handled, but I think inlining should be done when prescribed by the Inline=true annotation. As I understand, flags will be `-f -d=nonfScalarize,arrayConnect,combineSubscripts,evaluateAllParameters,vectorizeBindings --showStructuralAnnotations --frontendInline`, with `arrayConnect` being optional, depending on whether it works fine or not for the specific model at hand.

Bottom line, as I understand, with `--frontendInline` functions will be inlined by the NF, including operator record functions. Then, if `--newBackend` is activated, records will be preserved, while they will be flattened out if that flag is not activated.

@perost, do I miss something?

**Comment by casella** [comment]
I'll try this out as soon as there is a new nightly build available.

**Comment by perost** [comment]
I will take the silence as a good sign and close this, if there are more things that need to be fixed it's probably best to just open new issues anyway.

**Comment by casella** [comment]
Sorry, I should have closed this myself, the feature works as intented. Thanks!
